### ISSUE TITLE
COPY function to copy rows with nullable timestamp throws error

### ISSUE THREAD

**Issue by callum-p** [denishpatel/pg-clone-schema#81]
The below error is raised when using the COPY method on tables that have nullable timestamps:

```
ERROR:  Action: Copy Rows  Diagnostics: line=COPY yy, line 1, column archived_at: "N"
SQL statement "COPY sdf6.yy  FROM  '/tmp/cloneschema.tmp' (DELIMITER ',', NULL '');"
PL/pgSQL function clone_schema(text,text,boolean,boolean,boolean) line 1646 at EXECUTE. 22007. invalid input syntax for type timestamp with time zone: "N"
CONTEXT:  PL/pgSQL function clone_schema(text,text,boolean,boolean,boolean) line 1708 at RAISE
```

Table definition:

```
xx=# \d+ xx.yy;
                                                  Table "xx.yy"
      Column      |              Type              | Collation | Nullable |         Default          | Storage  | Stats target | Description 
------------------+--------------------------------+-----------+----------+--------------------------+----------+--------------+-------------
 id               | uuid                           |           | not null | public.gen_random_uuid() | plain    |              | 
 name             | text                           |           | not null |                          | extended |              | 
 publish_date     | timestamp(0) with time zone    |           | not null |                          | plain    |              | 
 start_date       | timestamp(0) with time zone    |           | not null |                          | plain    |              | 
 end_date         | timestamp(0) with time zone    |           | not null |                          | plain    |              | 
 archived_at      | timestamp(0) with time zone    |           |          |                          | plain    |              | 
 banner_color     | xx.banner_color |           | not null |                          | plain    |              | 
 undraw_banner_id | uuid                           |           | not null |                          | plain    |              | 
 created_at       | timestamp(0) with time zone    |           | not null |                          | plain    |              | 
 updated_at       | timestamp(0) with time zone    |           | not null |                          | plain    |              | 
 deleted_at       | timestamp(0) with time zone    |           |          |                          | plain    |              | 
 completed_at     | timestamp(0) with time zone    |           |          |                          | plain    |              | 

```

It does work when I update the clone-schema function to use this:

```
EXECUTE 'COPY ' || quote_ident(source_schema) || '.' || quote_ident(tblname) || ' TO ''/tmp/cloneschema.tmp'' WITH DELIMITER AS '','';';
EXECUTE 'COPY ' || quote_ident(dest_schema) || '.' || quote_ident(tblname) || ' FROM ''/tmp/cloneschema.tmp'' (DELIMITER '','');';
```

**Comment by MichaelDBA** [comment]
hmmm, I am not able to duplicate the error even after creating a table with null and not null timestamps, some rows with values for null timestamp columns and some rows without values for null timestamp columns.  It works fine for me.

What version of clone_schema are you using?  What postgres version?

**Comment by callum-p** [comment]
Hi @MichaelDBA thanks for taking a look. I'm away from the computer right now but I suspect it could be due to the row type "timestamp(0) with time zone". Did you have that specific type with your test?

I was using pg13. Not sure about minor version.  I'll get more specific info and some data to work with tomorrow.

**Comment by MichaelDBA** [comment]
Please make sure you are using the latest version.  When you run it, it will display the version at the top.

I still cannot duplicate your problem with the latest clone_schema on PG13.  Please try to run the below script to see if that works in your environment.

```
CREATE SCHEMA sample;
CREATE table sample.timestamptbl (akey int not null, avalue text, tmstmp timestamptz NOT NULL, tmstmp_null timestamptz NULL,  tmstmp_null2 timestamp(0) with time zone);
INSERT into sample.timestamptbl (akey, avalue, tmstmp, tmstmp_null, tmstmp_null2) VALUES (1,  'aaa', now(), now(), now());
INSERT into sample.timestamptbl (akey, avalue, tmstmp ) VALUES (1,  'aaa', now());
SELECT * from sample.timestamptbl;
 akey | avalue |            tmstmp             |          tmstmp_null          |      tmstmp_null2
------+--------+-------------------------------+-------------------------------+------------------------
    1 | aaa    | 2022-07-14 08:31:36.843239-04 | 2022-07-14 08:31:36.843239-04 | 2022-07-14 08:31:37-04
    1 | aaa    | 2022-07-14 08:32:18.932115-04 |                               |

SELECT clone_schema('sample', 'sample_clone1', true, false);
SELECT * from sample_clone1.timestamptbl;
 akey | avalue |            tmstmp             |          tmstmp_null          |      tmstmp_null2
------+--------+-------------------------------+-------------------------------+------------------------
    1 | aaa    | 2022-07-14 08:31:36.843239-04 | 2022-07-14 08:31:36.843239-04 | 2022-07-14 08:31:37-04
    1 | aaa    | 2022-07-14 08:32:18.932115-04 |                               |

```

**Comment by callum-p** [comment]
@MichaelDBA I believe the crucial peice of information I was missing from my original post is that the table contains a user defined data type which will cause the clone script to use COPY instead of INSERT INTO.

I was able to reproduce it with the latest version of the clone_schema function.

Steps:

```
CREATE SCHEMA sample;
CREATE TYPE sample.banner_color AS ENUM (
    'green',
    'blue',
    'lightblue',
    'purple',
    'red',
    'yellow',
    'orange',
    'grey',
    'pink'
);
CREATE table sample.timestamptbl (akey int not null, color sample.banner_color, avalue text, tmstmp timestamptz NOT NULL, tmstmp_null timestamptz NULL,  tmstmp_null2 timestamp(0) with time zone);
INSERT into sample.timestamptbl (akey, color, avalue, tmstmp, tmstmp_null, tmstmp_null2) VALUES (1,  'green', 'aaa', now(), now(), now());
INSERT into sample.timestamptbl (akey, color, avalue, tmstmp ) VALUES (1, 'red', 'aaa', now());


SELECT public.clone_schema('sample', 'sample_clone2', True, False, False);
```


```
Password for user test_user: 
psql (13.7 (Ubuntu 13.7-1.pgdg20.04+1), server 13.5)
SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, bits: 256, compression: off)
Type "help" for help.

db=# CREATE SCHEMA sample;
CREATE SCHEMA
db=# CREATE TYPE sample.banner_color AS ENUM (
db(#     'green',
db(#     'blue',
db(#     'lightblue',
db(#     'purple',
db(#     'red',
db(#     'yellow',
db(#     'orange',
db(#     'grey',
db(#     'pink'
db(# );
CREATE TYPE
db=# CREATE table sample.timestamptbl (akey int not null, color sample.banner_color, avalue text, tmstmp timestamptz NOT NULL, tmstmp_null timestamptz NULL,  tmstmp_null2 timestamp(0) with time zone);
CREATE TABLE                                                             ^
db=# INSERT into sample.timestamptbl (akey, color, avalue, tmstmp, tmstmp_null, tmstmp_null2) VALUES (1,  'green', 'aaa', now(), now(), now());
INSERT 0 1
db=# INSERT into sample.timestamptbl (akey, color, avalue, tmstmp ) VALUES (1, 'red', 'aaa', now());
INSERT 0 1
db=# 
-- ensuring I have the latest version of clone_schema function here
db=# \e
CREATE FUNCTION
NOTICE:  function public.clone_schema(text,text,pg_catalog.bool,pg_catalog.bool) does not exist, skipping
DROP FUNCTION
CREATE FUNCTION
db=# SELECT public.clone_schema('sample', 'sample_clone2', True, False, False);
NOTICE:  clone_schema version 1.7  June 29, 2022
INFO:  version: PostgreSQL 13.5 on x86_64-pc-linux-musl, compiled by gcc (Alpine 10.3.1_git20211027) 10.3.1 20211027, 64-bit
INFO:  OpSys = Linux
NOTICE:    COLLATIONS cloned:     0
NOTICE:       DOMAINS cloned:     0
NOTICE:         TYPES cloned:     3
NOTICE:     SEQUENCES cloned:     0
NOTICE:        TABLES cloned:     1
NOTICE:      SEQUENCES set:       0
NOTICE:     IDENTITIES set:       0
NOTICE:         VIEWS cloned:     0
NOTICE:     MAT VIEWS cloned:     0
NOTICE:     FUNCTIONS cloned:     0
NOTICE:      TRIGGERS cloned:     0
NOTICE:      RULES    cloned:     0
NOTICE:      POLICIES cloned:     0
NOTICE:   COMMENTS(1) cloned:     0
NOTICE:   COMMENTS(2) cloned:     0
NOTICE:    DFLT PRIVS cloned:     0
NOTICE:  SCHEMA PRIVS cloned:     0
NOTICE:    SEQ. PRIVS cloned:     0
NOTICE:    FUNC PRIVS cloned:     0
NOTICE:   TABLE PRIVS cloned:     7
ERROR:  Action: Copy Rows  Diagnostics: line=COPY timestamptbl, line 2, column tmstmp_null: "N"
SQL statement "COPY sample_clone2.timestamptbl  FROM  '/tmp/cloneschema.tmp' (DELIMITER ',', NULL '');"
PL/pgSQL function clone_schema(text,text,boolean,boolean,boolean) line 1644 at EXECUTE. 22007. invalid input syntax for type timestamp with time zone: "N"
CONTEXT:  PL/pgSQL function clone_schema(text,text,boolean,boolean,boolean) line 1706 at RAISE

```

**Comment by callum-p** [comment]
I can confirm that it works. Thanks a lot!
